### PR TITLE
feat(number-input): number-input with @carbon/react sync

### PR DESCRIPTION
**Pull request by annawen1**
### Related Ticket(s)

[number-input]: sync with @carbon/react version #10111

### Description

@carbon/react syncing with `number-input` component. This involves changes to the base `input` component as well.

@carbon/react version: https://react.carbondesignsystem.com/?path=/story/components-numberinput--playground

### Changelog

**New**

- new styles with addition of warning icon, different icons for the increment/decrement buttons
- new functionality - readonly, warn, hide steppers, hide label, etc

### CODE REVIEW

**Review by RichKummer — COMMENTED** [review]
This is looking good! Just a few things things I saw related to `warn`: 

Looking at the react version, there should be a divider added when the `warn` icon appears. See the react version here:
Screen Shot 2023-03-16 at 12 00 19 PM

In the react version, when both `warn` _and_ `readonly` are true, the warn icon is hidden from view. In this PR, the warn icon still appears when both are true. This is the react build with both set to true:
Screen Shot 2023-03-16 at 12 01 55 PM

File: packages/carbon-web-components/src/components/number-input/number-input.ts [hunk 1]
```diff
@@ -314,42 +307,48 @@ class BXNumberInput extends BXInput {
...
+      normalizedProps.icon = invalidIcon;
+      normalizedProps['slot-name'] = 'invalid-text';
+      normalizedProps['slot-text'] = this.invalidText;
+    } else if (normalizedProps.warn) {
+      normalizedProps.icon = warnIcon;
+      normalizedProps['slot-name'] = 'warn-text';
+      normalizedProps['slot-text'] = this.warnText;
+    }
+
+    console.log('normalizeProps', normalizedProps);
```
Comment (ariellalgilmore) [inline packages/carbon-web-components/src/components/number-input/number-input.ts:320]:
```suggestion
```

File: packages/carbon-web-components/src/components/number-input/stories/helpers.ts [hunk 2]
```diff
@@ -1,44 +1,48 @@
...
   step: number('Value to step the input by (step)', 1),
-  size: select('Input size (size)', sizes, INPUT_SIZE.REGULAR),
-  placeholder: textNullable(
-    'Placeholder text (placeholder)',
-    'Optional placeholder text'
-  ),
-  invalid: boolean('Invalid (invalid)', false),
+  size: select('Size (size)', sizes, INPUT_SIZE.MEDIUM),
   onInput: action('input'),
   mobile: boolean('Mobile mode (mobile)', false),
```
Comment (ariellalgilmore) [inline packages/carbon-web-components/src/components/number-input/stories/helpers.ts:45]: is this still necessary for web components? https://github.com/carbon-design-system/carbon/blob/9d0a08fc5ccaf7259f256545b172cbf4dee8991a/docs/migration/v11.md?plain=1#L929

Comment (annawen1) [reply packages/carbon-web-components/src/components/number-input/stories/helpers.ts:45]: oh good catch - looks like it doesn't even have mobile specific styles anymore - will remove!

**Review by ariellalgilmore — APPROVED** [review]
LGTM!

**Comment by RichKummer** [comment]
Looks like the `warn` and `readonly` issues were resolved! I did find that unlike the react build, the border next to `warn` disappears when the Number input is `disabled`.

Screen Shot 2023-03-22 at 4 05 04 PM

The rest is looking good!

**Review by RichKummer — APPROVED** [review]
Looks great, thanks for the fixes @annawen1 ! 🚀

**Pull request merged**